### DISCUSSION TITLE
How to deal with a "many to many"(many2many) relationship from a RDB in our CSV2RDF work?

### DISCUSSION THREAD

**Discussion by candlecao** [DDMAL/linkedmusic-datalake#152] [Ideas]
How to deal with a "many to many" relationship from a RDB in our CSV2RDF work?

**Comment by candlecao** [comment]
A "many to many" relationship is a term within the scope of mapping cardinality, also of schema designing as to an relational database(RDB). Such as one album has many tracks and otherwise one track can appear in more than one albums. This is a quite common phenomenon. For many-to-many relationships, It's determined to use a "relation set" spreadsheet for modeling instead of an "entity set" spreadsheet. So, in our process of CSV2RDF, we should distinguish between "entity set" and "relation set" if possible. 

In views of schema of an RDB, the fundamental difference between "entity set" and "relation set" can be seen from the composition of primary key of a spread sheet(or CSV table):
> 1. For an entity set, We usually render the primary using a single attribute on the first column of the table.
> 2. For an relation set, it often has a composite primary key, typically formed by combining the primary keys of the participating entity sets, thereby uniquely identifying each relationship instance.

The significance of a relation set is, to contain the attributes that describe the relation, so in perspective of a triple of subject-predicate-object, **the attributes of a relation set are like an expression of "adverbial"...**

### Let's take some from TheSession DataBase for example:<br>

1. The dumped CSVs don't align with the original schema, I supposed. (I mean, it probably doesn't differentiate the tables the way the tables are organized in the database.)
2. So we can find examples like:
> 2.1 in recordings.csv
<html xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40">
<head>

<meta name=Generator content="Microsoft Excel">

<style>

</style>
</head>
<body>


recording_id | artist | recording | track | number | tune | tune_id | artist_url
-- | -- | -- | -- | -- | -- | -- | --
https://thesession.org/recordings/1032 | Various Artists | Music For The Sets Volume 3: Pay The Piper | 12 | 1 | Cooley's | https://thesession.org/tunes/1 | https://thesession.org/recordings/artists/39
https://thesession.org/recordings/1032 | Various Artists | Music For The Sets Volume 3: Pay The Piper | 12 | 2 | Bird In The Bush, The | https://thesession.org/tunes/629 | https://thesession.org/recordings/artists/39
https://thesession.org/recordings/3536 | Accordéonistes Du Québec | Dans Tous Les Cantons | 23 | 1 | Cooley's | https://thesession.org/tunes/1 | https://thesession.org/recordings/artists/2004



</body>

</html>

_--we can see `<https://thesession.org/recordings/1032>` has 2 "tune_id"s and `<https://thesession.org/tunes/1>` appears in 2 "recording_id"s._

In this table, the `track` column and the `number` column are attributes of the "recording-tune" relationship instead of attributes solely of `recording_id` only. The `track` value indicates the track NO. of the tune within the recording; the `number` value represents the position or sequence of the tune in the recording or possibly something else. Hence we have to use named graph or blank node.(see the following)

> 2.2 in sets.csv
<html xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40">
<head>

<meta name=Generator content="Microsoft Excel">

<style>

</style>
</head>
<body>


tuneset_id | username | name | tune_id
-- | -- | -- | --
https://thesession.org/members/1/sets/5 | Jeremy | Cooley's | https://thesession.org/tunes/1
https://thesession.org/members/1/sets/5 | Jeremy | Wise Maid, The | https://thesession.org/tunes/118
https://thesession.org/members/46273/sets/22 | michaelkeogh | Cooley's | https://thesession.org/tunes/1



</body>

</html>

_--we can see `<https://thesession.org/members/1/sets/5>` has 2 "tune_id"s and `<https://thesession.org/tunes/1>` appears in 2 "tuneset_id"s._

**Reply by fujinaga** [reply]
Can you give us the example from The Session?

**Reply by candlecao** [reply]
Yes. I am doing that right now.

**Comment by candlecao** [comment]
### To render "many to many" relationship with 2 methods:
Let us convert the 3 rows of data from aforementioned excerpt of "recordings.csv" into(only choose columns--recording_id, track, number, tune_id):
#### 1. Named Graph
```
@prefix wdt: <http://www.wikidata.org/prop/direct/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<https://thesession.org/recordings/1032> <https://thesession.org/tunes> <https://thesession.org/tunes/1> <urn:relationship:1032_1> .
<https://thesession.org/tunes/1> wdt:P658 "12"^^xsd:int <urn:relationship:1032_1> .
<https://thesession.org/tunes/1> wdt:P1114 "1"^^xsd:int <urn:relationship:1032_1> .

<https://thesession.org/recordings/1032> <https://thesession.org/tunes> <https://thesession.org/tunes/629> <urn:relationship:1032_629> .
<https://thesession.org/tunes/629> wdt:P658 "12"^^xsd:int <urn:relationship:1032_629> .
<https://thesession.org/tunes/629> wdt:P1114 "2"^^xsd:int <urn:relationship:1032_629> .

<https://thesession.org/recordings/3536> <https://thesession.org/tunes> <https://thesession.org/tunes/1> <urn:relationship:3536_1> .
<https://thesession.org/tunes/1> wdt:P658 "23"^^xsd:int <urn:relationship:3536_1> .
<https://thesession.org/recordings/3536> <https://thesession.org/tunes/1> wdt:P1114 "1"^^xsd:int <urn:relationship:3536_1> .                         
```

If you upload it on Virtuoso, then you can query in SPARQL like this:
```
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
SELECT * where {
  graph ?g {
    ?s ?p ?o
  }
}
```
or refer to <https://github.com/DDMAL/linkedmusic-datalake/issues/98> (which however might be a hallucination)
#### 2. Blank node
```
@prefix wdt: <http://www.wikidata.org/prop/direct/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
<https://thesession.org/recordings/1032> <https://thesession.org/tunes> [
  <https://thesession.org/tunes> <https://thesession.org/tunes/1>;
 wdt:P658 "12"^^xsd:int;  
 wdt:P1114 "1"^^xsd:int
].

<https://thesession.org/recordings/1032> <https://thesession.org/tunes> [
  <https://thesession.org/tunes> <https://thesession.org/tunes/629>;
 wdt:P658 "12"^^xsd:int;  
 wdt:P1114 "2"^^xsd:int
].

<https://thesession.org/recordings/3536> <https://thesession.org/tunes> [
  <https://thesession.org/tunes> <https://thesession.org/tunes/1>;
 wdt:P658 "23"^^xsd:int;  
 wdt:P1114 "1"^^xsd:int
].
```

### 
(That will directly impact our programing for CSV2RDF into Virtuoso.)

**Comment by candlecao** [comment]
Hi Ich(@fujinaga ), it's so weird that I cannot upload the Named Graph content to Virtuoso (by putting them in a file with suffix .nq) But if I don't use the `@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .`, it then can be uploaded. Like this:
```
<https://thesession.org/recordings/1032> <http://example.org/includes> <https://thesession.org/tunes/1> <urn:relationship:1032_1> .
<https://thesession.org/recordings/1032> <http://www.wikidata.org/prop/direct/P658> "12"^^<http://www.w3.org/2001/XMLSchema#int> <urn:relationship:1032_1> .
<https://thesession.org/tunes/1> <http://www.wikidata.org/prop/direct/P1114> "1"^^<http://www.w3.org/2001/XMLSchema#int> <urn:relationship:1032_1> .
#...
```
Do you know why?

**Comment by candlecao** [comment]
Any way, I put up the issue on https://github.com/openlink/virtuoso-opensource/issues/1313.

**Comment by candlecao** [comment]
We summarized 4 methods:<br>
e.g.:
> I have a databases with a list of courses, a list of students taking these courses, and the grades that each student received in each of the courses. How can I structure this in RDF (Turtle)?

1. Blank nodes

2. Named Graphs

3. Reification: refer to A Semantic Web Primer [(Greece) Grigoris Antoniou, The MIT Press, 2012] > 2.2.5 Pointing to Statements and Graphs
> One is called reification. The key idea behind reification is to introduce an auxiliary object, say, LocationStatement, and relate it to each of the three parts of the original statement through the properties subject, predicate, and object. In the preceding example the subject of LocationStatement would be BaronWayBuilding, the predicate would be location, and the object Amsterdam. We can then refer to the statement in the subject of another triple that defines the creator. Figure 2.4 depicts the resulting graph. Again, the full URIs are not shown for presentation purposes.
<img width="748" alt="image" src="https://github.com/user-attachments/assets/5c213fc7-9fe1-4eec-b547-fdbb30dddf05">


4. "Concrete nodes"
> This method is similar to "Blank nodes". For example, we render the grade as a concrete node which have predicates connecting the students, courses and specific grades. See https://chatgpt.com/share/a87857dd-8b76-4be4-a6ff-2503ef5b7c22.
